src/zarr/storage/_cache.py

```diff
@@ -0,0 +1,698 @@
+    --------
+    The example below wraps a LocalStore with an LRU cache for demonstration::
+
+        >>> import tempfile
+        >>> import zarr
+        >>> from zarr.storage import LocalStore
+        >>>
+        >>> # Create a temporary directory for the example
+        >>> temp_dir = tempfile.mkdtemp()
+        >>> store = LocalStore(temp_dir)
+        >>>
+        >>> # Create some test data first
+        >>> arr = zarr.create((1000, 1000), chunks=(100, 100), store=store, dtype='f4')
+        >>> arr[:] = 42.0
+        >>>
+        >>> # Now wrap with cache for faster access
+        >>> cached_store = zarr.LRUStoreCache(store, max_size=1024 * 1024 * 256)  # 256MB cache
+        >>> cached_arr = zarr.open(cached_store)
+        >>>
+        >>> # First access loads from disk and caches
+        >>> data1 = cached_arr[0:100, 0:100]  # Cache miss
+        >>>
+        >>> # Second access uses cache (much faster for remote stores)
+        >>> data2 = cached_arr[0:100, 0:100]  # Cache hit
+
+    For remote stores where the performance benefit is more apparent::
+
+        >>> from zarr.storage import RemoteStore
+        >>> # remote_store = RemoteStore.from_url("https://example.com/data.zarr")
+        >>> # cached_remote = zarr.LRUStoreCache(remote_store, max_size=2**28)
+
+
+    """
+
+    supports_writes: bool = True
+    supports_deletes: bool = True
+    supports_partial_writes: bool = True
+    supports_listing: bool = True
+
+    root: Path
+
+    def __init__(self, store: Store, *, max_size: int) -> None:
+        if max_size <= 0:
+            raise ValueError("max_size must be a positive integer (bytes)")
+
+        # Always inherit read_only state from the underlying store
+        read_only = getattr(store, "read_only", False)
+
+        # Call parent constructor with read_only parameter
+        super().__init__(read_only=read_only)
+
+        self._store = store
+        self._max_size = max_size
+        self._current_size = 0
+        self._keys_cache: list[str] | None = None
+        self._contains_cache: dict[Any, Any] = {}
+        self._listdir_cache: dict[str | None, list[str]] = {}
+        self._values_cache: OrderedDict[str, Any] = OrderedDict()
+        self._mutex = Lock()
+        self.hits = self.misses = 0
+
+        # Handle root attribute if present in underlying store
+        if hasattr(store, "root"):
+            self.root = store.root
+        else:
+            self.root = Path("/")  # Default root path
+
+    @classmethod
+    async def open(cls, store: Store, *, max_size: int, **kwargs: Any) -> "LRUStoreCache":
+        """
+        Create and open the LRU cache store.
+
+        Parameters
+        ----------
+        store : Store
+            The underlying store to wrap with caching.
+        max_size : int
+            The maximum size that the cache may grow to, in number of bytes.
+        **kwargs : Any
+            Additional keyword arguments passed to the store constructor.
```

Comment:
Because `max_size` is explicitly  in the signature to `open()`, which is the only keyword argument to `__init__`, `kwargs` don't need to be added to the signature.